src/storage/hnsw_usearch.cpp

```diff
@@ -0,0 +1,500 @@
+#include "boot_sr.h"
+#include "file_io.h"
+#include "system_parameter.h"
+#include "dbtype.h"
+#include "porting.h"
+#include "vector_distance_enum.h"
+#include <fstream>
+#include <filesystem>
+
+#include "usearch/index.hpp"
+#include "usearch/index_dense.hpp"
+#include "usearch/index_plugins.hpp"
+
+// XXX: SHOULD BE THE LAST INCLUDE HEADER
+#include "memory_wrapper.hpp"
+
+
+#include "strict_warnings_on.hpp"// TODO : When cub_server terminates, hnsw_index_id will be reset to 0.
+//        This is not a problem in current implementation, but it may be a problem in the future,
+//        such as duplicate hnsw_index_id when cub_server restarts.
+//        We need to consider a better way to identify the hnsw index.
+
+
+using namespace unum;
+
+int hnsw_index_id = 0;
+std::unordered_map<int, std::unique_ptr<usearch::index_dense_t>> hnsw_index_map;
+char hnsw_index_directory[PATH_MAX] = {0};
+bool hnsw_index_directory_created = false;
+static std::mutex hnsw_elem_mutex;
+
+static int dump_hnsw_index (int hnsw_id, const std::unique_ptr<usearch::index_dense_t> &index);
+static int get_hnsw_index_file_path (int hnsw_id, char *out_path);
+static int create_hnsw_index_directory ();
+static bool is_hnsw_index_file_exists (int hnsw_id);
+static int load_hnsw_index_from_file (int hnsw_id);
+static int hnsw_check_and_load_index (int hnsw_id);
+static void get_new_hnsw_index_id ();
+static int64_t encode_oid (const OID &oid);
+static OID decode_oid (int64_t encoded_oid);
+
+BTID *
+xhnsw_add_index (THREAD_ENTRY *thread_p, BTID *btid, int dimension = 10, int hnsw_M = 16, int hnsw_efConstruction = 64,
+		 int metric = DB_VECTOR_DISTANCE_METRIC::METRIC_EUCLIDEAN)
+{
+  usearch::metric_kind_t metric_kind = usearch::metric_kind_t::unknown_k;
+  switch (metric)
+    {
+    case METRIC_UNKNOWN:
+    case METRIC_COSINE:
+      metric_kind = usearch::metric_kind_t::cos_k;
+      break;
+    case METRIC_DOT:
+      // TODO (CUBVEC): Faiss does not support cosine distance directly?
+      metric_kind = usearch::metric_kind_t::ip_k;
+      break;
+
+    case METRIC_EUCLIDEAN:
+      metric_kind = usearch::metric_kind_t::l2sq_k;
+      break;
+
+    case METRIC_MANHATTAN:
+      // unsupported metric
+      ASSERT_CUBVEC (false);
+      break;
+
+    default:
+      ASSERT_CUBVEC (false);
+    }
+
+  get_new_hnsw_index_id ();
+
+  usearch::metric_punned_t metric_punned (static_cast <std::size_t> (dimension), metric_kind,
+					  usearch::scalar_kind_t::f32_k);
+
+  usearch::index_dense_config_t config;
+  config.connectivity = hnsw_M;
+  config.expansion_add = hnsw_efConstruction;
+
+  auto index_ptr = std::make_unique<usearch::index_dense_t> (
+			   usearch::index_dense_t::make (metric, config)
+		   );
+
+  index_ptr->reserve (10000);
+
+  hnsw_index_map[hnsw_index_id] = std::move (index_ptr);
+
+  btid->vfid.volid = -1;
+  btid->vfid.fileid = -1;
+  btid->root_pageid = hnsw_index_id;
+
+  er_log_debug (ARG_FILE_LINE, "HNSW Index added with ID %d", hnsw_index_id);
```

Comment:
디버깅성 로그라면, hnsw_usearch.cpp 내 모두 _er_log_debug() 사용 권장합니다.

```suggestion
  _er_log_debug (ARG_FILE_LINE, "HNSW Index added with ID %d", hnsw_index_id);
```